sql: introduce CCL-only statements: `GRANT|REVOKE <role>`

mberhault opened the pull request:
Release note (sql change): introduce experimental statements
`GRANT|REVOKE <role>` to grant/revoke role memberships.

Part of [role-based access control](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20171220_sql_role_based_access_control.md).

`GRANT <roles> TO <users or roles> [WITH ADMIN OPTION]` adds `users or roles` as members of `roles` or adds the ADMIN
OPTION if they already exist.

`REVOKE [ADMIN OPTION FOR] <roles> FROM <users or roles>` removes the
ADMIN OPTION or the entire membership.

`ADMIN OPTION` allows the member to modify this role's memberships, it is
inherited. It can be added/removed for a given membership if the
relationship exists.

The help message for `GRANT` and `REVOKE` cover both privileges and roles as we don't have a good way to split them.

See [RFC](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20171220_sql_role_based_access_control.md#grant-role) for behavior details.

Commit:
sql: introduce CCL-only statements: `GRANT|REVOKE <role>`

Release note (sql change): introduce experimental statements
`GRANT|REVOKE <role>` to grant/revoke role memberships.

Part of [role-based access
control](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20171220_sql_role_based_access_control.md).

`GRANT <roles> TO <users or roles> [WITH ADMIN OPTION]` adds `users or roles` as members of `roles` or adds the ADMIN
OPTION if they already exist.

`REVOKE [ADMIN OPTION FOR] <roles> FROM <users or roles>` removes the
ADMIN OPTION or the entire membership.

`ADMIN OPTION` allows the member to modify this role's memberships, it is
inherited. It can be added/removed for a given membership if the
relationship exists.

The help message for `GRANT` and `REVOKE` cover both privileges and roles as we don't have a good way to
split them.

See
[RFC](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20171220_sql_role_based_access_control.md#grant-role) for behavior details.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21341)
<!-- Reviewable:end -->

jordanlewis commented:
Reviewed 20 of 20 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/role.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2Qfh-zDWBCzJMsVqBR:-L2Qfh-zDWBCzJMsVqBS:bg07d8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L82)):*
> ```Go
> 		for _, r := range grant.Roles {
> 			if isAdmin, ok := allRoles[string(r)]; !ok || !isAdmin {
> 				return nil, errors.Errorf("%s is not a superuser or role admin for role %s", p.User(), r)
> ```

All of these user-facing errors in this file need pgerror codes. Try the equivalent in postgres with `\set VERBOSITY verbose` to see the error code we should match, then add it to the errors with the `pgerror.NewErrorf(pgerror.CodeFoo...`

---

*[pkg/ccl/sqlccl/role.go, line 98 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QflhtKSZvArEk2YiB:-L2QflhtKSZvArEk2YiC:bp2ccss) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L98)):*
> ```Go
> 	for _, r := range grant.Roles {
> 		if isRole, ok := users[string(r)]; !ok || !isRole {
> 			return nil, errors.Errorf("role %s does not exist", r)
> ```

pgerror

---

*[pkg/ccl/sqlccl/role.go, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QfmShJ2mu1LT2yqiG:-L2QfmShJ2mu1LT2yqiH:bp2ccss) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L105)):*
> ```Go
> 	for _, m := range grant.Members {
> 		if _, ok := users[string(m)]; !ok {
> 			return nil, errors.Errorf("user or role %s does not exist", m)
> ```

pgerror

---

*[pkg/ccl/sqlccl/role.go, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QfoQjf0rGVxUL60cY:-L2QfoQjf0rGVxUL60cZ:bp2ccss) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L132)):*
> ```Go
> 				// self-cycle.
> 				return nil, errors.Errorf("%s cannot be a member of itself", m)
> 			}
> ```

pgerror

---

*[pkg/ccl/sqlccl/role.go, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QfEELuEHutLEXRI8l:-L2QfEELuEHutLEXRI8m:bp2ccss) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L136)):*
> ```Go
> 			if memberOf, ok := allRoleMemberships[r]; ok {
> 				if _, ok = memberOf[m]; ok {
> 					return nil, errors.Errorf("making %s a member of %s would create a cycle", m, r)
> ```

pgerror

---

*[pkg/sql/authorization.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QcFDRZjysTJeb0NvX:-L2QcFDRZjysTJeb0NvY:bun61h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/authorization.go#L101)):*
> ```Go
> 	toVisit := []string{member}
> 
> 	// Placeholders don't work with arrays, so let's do one by one.
> ```

What do you mean? That sounds like a bug. Also, how would arrays help here?

---

*[pkg/sql/user.go, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2Qctyf73g8Bw3kGY_A:-L2QctygGJw0zRT7u2eW:b-soj9zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/user.go#L64)):*
> ```Go
> func (p *planner) GetAllUsersAndRoles(ctx context.Context) (map[string]bool, error) {
> 	query := `SELECT username,"isRole"  FROM system.users`
> 	newPlanner := makeInternalPlanner("get-all-users", p.txn, security.RootUser, p.session.memMetrics)
> ```

should this be changed to `get-all-users-and-roles`?

---

*[pkg/sql/parser/sql.y, line 7515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QdkpOo1TEZDwRUuxC:-L2QdkpOo1TEZDwRUuxD:bfzzgz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/parser/sql.y#L7515)):*
> ```Yacc
> | COLLATION
> | CROSS
> | DELETE
> ```

Hmm, definitely not great that we had to do this - it means that insert/delete/update/drop can't be used as column names anymore. Postgres's grammar didn't have to. What mistake are we making? ADMIN is in unsreserved_keyword in postgres as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/role.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2Qfh-zDWBCzJMsVqBR:-L2Qhpxhbump7Z8eFa44:b-kothv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L82)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

All of these user-facing errors in this file need pgerror codes. Try the equivalent in postgres with `\set VERBOSITY verbose` to see the error code we should match, then add it to the errors with the `pgerror.NewErrorf(pgerror.CodeFoo...`
</blockquote></details>

ugh. fun. I guess I'll spend some time finding the same errors.  We have a lot of other places without those in the user/role/privilege code paths.

---

*[pkg/sql/authorization.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QcFDRZjysTJeb0NvX:-L2QiHbS1pAgW6BvW4fm:b-vuf9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/authorization.go#L101)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What do you mean? That sounds like a bug. Also, how would arrays help here?
</blockquote></details>

since I may have more than one thing to expand, I wanted to do `WHERE "member" = ANY($1)` where `$1` is the `toVisit` array. Didn't work though, it seems we don't support array placeholders.

---

*[pkg/sql/user.go, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2Qctyf73g8Bw3kGY_A:-L2QiaGVPCqxYAsJNIQa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/user.go#L64)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

should this be changed to `get-all-users-and-roles`?
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 7515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QdkpOo1TEZDwRUuxC:-L2Qib2zWhqufAqtSzKg:bb28vo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/parser/sql.y#L7515)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm, definitely not great that we had to do this - it means that insert/delete/update/drop can't be used as column names anymore. Postgres's grammar didn't have to. What mistake are we making? ADMIN is in unsreserved_keyword in postgres as well.
</blockquote></details>

Sure they can, you just need to quote the column name. For such keywords, that doesn't seem too crazy. I'd be happy to see those demoted to less severe keywords, but I'm not sure how.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 17 of 20 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/ccl/sqlccl/role.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2Qfh-zDWBCzJMsVqBR:-L2QvPgRdUMuyP54ZKwq:b-u2nff6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L82)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

ugh. fun. I guess I'll spend some time finding the same errors.  We have a lot of other places without those in the user/role/privilege code paths.
</blockquote></details>

Done for `sqlccl/roles.go`. I'll file a bug to audit errors codes in others places. I'll start with myself for user/role/privileges paths.

---

*[pkg/ccl/sqlccl/role.go, line 98 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QflhtKSZvArEk2YiB:-L2QvK0IEDxE5w0EkM1y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L98)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

pgerror
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/role.go, line 105 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QfmShJ2mu1LT2yqiG:-L2QvKSiIn4X6Nj1JFmJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L105)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

pgerror
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/role.go, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QfoQjf0rGVxUL60cY:-L2QvLLOynYyZ_9RUdK5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L132)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

pgerror
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/role.go, line 136 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QfEELuEHutLEXRI8l:-L2QvLfjEiGSJRqArgS_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/ccl/sqlccl/role.go#L136)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

pgerror
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
All comments addressed. PTAL.

jordanlewis commented:
Review status: 17 of 20 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/authorization.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QcFDRZjysTJeb0NvX:-L2R381WLCvFamLRhz2n:bqxjyn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/authorization.go#L101)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

since I may have more than one thing to expand, I wanted to do `WHERE "member" = ANY($1)` where `$1` is the `toVisit` array. Didn't work though, it seems we don't support array placeholders.
</blockquote></details>

I believe we do actually support array placeholders, at least via pgwire. Perhaps the thing that's missing is teaching `QueryRowsInTransaction` to send the right kind of object. For Go libpq I believe you need to turn the array into some kind of Postgres-specific array type. In other words I think we could get this working without too much trouble but yeah doesn't seem necessary for this PR - just something to keep in mind.

---

*[pkg/sql/parser/sql.y, line 7515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QdkpOo1TEZDwRUuxC:-L2R3aIgse-ggz2gndDt:bo5wl8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/parser/sql.y#L7515)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Sure they can, you just need to quote the column name. For such keywords, that doesn't seem too crazy. I'd be happy to see those demoted to less severe keywords, but I'm not sure how.
</blockquote></details>

I don't know how to do it at face value either, but I'm convinced that it's possible because Postgres's grammar supports these options without having to restrict these keywords. I don't want to merge this as-is because it could break user queries.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 17 of 20 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/authorization.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QcFDRZjysTJeb0NvX:-L2R4EzXmgCPsEQaLBQJ:bh60ebr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/authorization.go#L101)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I believe we do actually support array placeholders, at least via pgwire. Perhaps the thing that's missing is teaching `QueryRowsInTransaction` to send the right kind of object. For Go libpq I believe you need to turn the array into some kind of Postgres-specific array type. In other words I think we could get this working without too much trouble but yeah doesn't seem necessary for this PR - just something to keep in mind.
</blockquote></details>

Ok, let's leave it as is. This isn't a part that needs to be optimized anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 17 of 20 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/sql.y, line 7515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QdkpOo1TEZDwRUuxC:-L2R4LLwCeEfNdTuK47Y:bxgy7ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/parser/sql.y#L7515)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't know how to do it at face value either, but I'm convinced that it's possible because Postgres's grammar supports these options without having to restrict these keywords. I don't want to merge this as-is because it could break user queries.
</blockquote></details>

I guess we need to bring some other people in. Will ping the sql channel.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
If we don't want to promote privilege names, we'll need to do what postgres does and make them just plain strings parsed later. We probably also need a proper role/user definition somewhere.

I'll leave this here for now and get back to it later, this is just dumb.

mberhault commented:
Done in the second commit. The privilege list is now a `[]string` parsed at execution time. It is used by both `GRANT <roles>` and `GRANT <privileges>`. This allows us to demote `CREATE` and `GRANT` a bit.

mberhault commented:
Review status: 14 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/authorization.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QcFDRZjysTJeb0NvX:-L2S3pGvZfBRbbDQFSQI:bd771d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/authorization.go#L101)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Ok, let's leave it as is. This isn't a part that needs to be optimized anyway.
</blockquote></details>

Done. Removed comment.

---

*[pkg/sql/parser/sql.y, line 7515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2QdkpOo1TEZDwRUuxC:-L2S3r7i9z91QafZafU3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb666cd98d4aa61f27d126c12323b295b32260b/pkg/sql/parser/sql.y#L7515)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I guess we need to bring some other people in. Will ping the sql channel.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented on pkg/cmd/docgen/diagrams.go:
```diff
@@ -500,7 +500,7 @@ var specs = []stmtSpec{
 	{
 		name:    "grant_stmt",
-		inline:  []string{"privileges", "privilege_list", "privilege", "targets", "grantee_list", "table_pattern_list", "name_list"},
+		inline:  []string{"privileges", "privilege_list", "privilege", "targets", "table_pattern_list", "name_list"},
```
@mjibson: is there a process for keeping this up to date? A missing type will cause a crash (it did here when regenerating the docs), but what if we forget to add a new one?

mberhault commented:
Ping?

jordanlewis commented:
Looks like some of the commits were rebased to the wrong parent? I think the parsing at execution time ended up at r7 instead of r6, unless reviewable  is misleading me.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for one more grammar issue.

---

Reviewed 1 of 3 files at r2, 16 of 16 files at r5, 5 of 5 files at r6, 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[docs/generated/sql/bnf/stmt_block.bnf, line 1263 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2_VQQMVtzTKg7yCjb5:-L2_VQQMVtzTKg7yCjb6:b-qveq7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/55ba117e369807a64000b19307b8e30f210d61c4/docs/generated/sql/bnf/stmt_block.bnf#L1263)):*
> ```
> type_func_name_keyword ::=
> 	'COLLATION'
> 	| 'CREATE'
> ```

This also seems unnecessary and is a regression from the current state of things. It should stay in the normal keywords section. Am I going overboard? I'm just concerned with compatibility.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-:-L2_Vf2vuqkwm4RopfVh:b-bexg61)*
<!-- Sent from Reviewable.io -->

mberhault commented:
I honestly can't tell what reviewable does with revisions. I rebased after pulling master and fixed conflicts, then had to fix my fixes to the conflicts. 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[docs/generated/sql/bnf/stmt_block.bnf, line 1263 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2_VQQMVtzTKg7yCjb5:-L2_XMGFCSonK2ZTRQq-:b-bm26mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/55ba117e369807a64000b19307b8e30f210d61c4/docs/generated/sql/bnf/stmt_block.bnf#L1263)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This also seems unnecessary and is a regression from the current state of things. It should stay in the normal keywords section. Am I going overboard? I'm just concerned with compatibility.
</blockquote></details>

I'm happy leaving `GRANT` and `CREATE` in reserved keywords. I don't think demoting them hurts anything though, unless it means we have to promote them again soon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-:-L2_X8ycDkLn6M2UcJMX:b8siqbr)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Both `GRANT` and `CREATE` are in `reserved_keyword` in postgres' grammar which is probably how ours ended up there. I'm happy moving them back.

knz commented:
LGTM with nit on the contextual help strings.

---

Reviewed 7 of 20 files at r1, 1 of 3 files at r2, 3 of 16 files at r5, 2 of 5 files at r6, 7 of 7 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/sql.y, line 2118 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2_tPerh7QgknlXjDZA:-L2_tPesh-Pgj2lFttuG:bmtq930) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb064f015d22bf36d8eaa75f57b26058c4e4cba/pkg/sql/parser/sql.y#L2118)):*
> ```Yacc
> //   GRANT {ALL | <privileges...> } ON <targets...> TO <grantees...>
> // Grant role membership (CCL only):
> //   GRANT <roles> TO <roles or users> [WITH ADMIN OPTION]
> ```

`<roles>` -> `<roles...>` 
`<roles or users>` -> `<grantees...>`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-:-L2_tjtLnZM81QeGNhWw:b-ypxury)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 18 of 22 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/sql.y, line 2118 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L2_tPerh7QgknlXjDZA:-L2_vy7ns3GzE8tJLIL_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb064f015d22bf36d8eaa75f57b26058c4e4cba/pkg/sql/parser/sql.y#L2118)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`<roles>` -> `<roles...>` 
`<roles or users>` -> `<grantees...>`


</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 18 of 22 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/authorization.go, line 118 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341#-L71hUR-4MVG3I2nic21:-L71hUR-4MVG3I2nic22:b6gk1er) ([raw file](https://github.com/cockroachdb/cockroach/blob/4572dc8dd5ecec8c19a9daef22666439acc2467e/pkg/sql/authorization.go#L118)):*
> ```Go
> 			ctx,
> 			"expand-roles",
> 			p.Txn(),
> ```

is it OK that this is done in `p.Txn()`, as opposed to a fresh txn?
Besides making me wonder why (the default should be a fresh txn), what happens if a membership is revoked but then I run an AsOfSysTime txn and I get to live forever?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21341)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Good catch, this is fixed in #24284
